ddbToEsDlqLambda/index.ts

```diff
@@ -0,0 +1,151 @@
+        });
+
+    // resp.Successful contains successfully deleted messages.
+    if (resp.Successful) {
+        logger.debug('Deleted messages', JSON.stringify(resp.Successful));
+    }
+
+    // resp.Failed contains messages that were failed to be deleted
+    if (resp.Failed && resp.Failed.length) {
+        throw new Error(`Failed to delete messages: ${JSON.stringify(resp.Failed)}`);
+    }
+};
+
+const invokeDdbToEsLambda = async (records) => {
+    const resp = await lambda
+        .invoke({
+            FunctionName: DDB_TO_ES_LAMBDA_NAME,
+            InvocationType: INVOCATION_TYPE,
+            Payload: JSON.stringify(records),
+        })
+        .promise()
+        .catch((e) => {
+            throw new Error(`Failed to invoke DdbToEsLambda: ${JSON.stringify(e)}`);
+        });
+
+    // StatusCode 200 only means successful invocation.
+    if (resp.StatusCode >= 400) {
+        throw new Error(`Failed to invoke DdbToEsLambda: ${JSON.stringify(resp.Payload)}`);
+    }
+
+    // When errors occur during execution, StatusCode is 200, but FunctionError is present.
+    if (resp.FunctionError) {
+        throw new Error(`Failed to invoke DdbToEsLambda: ${JSON.stringify(resp.Payload)}`);
+    }
+
+    logger.debug('Invoked DdbToEsLambda with records', JSON.stringify(records));
+};
+
+const getRecordsFromDdbStream = async (message) => {
+    const resp = await dynamodbstreams
+        .getShardIterator({
+            ShardId: message.DDBStreamBatchInfo.shardId,
+            ShardIteratorType: 'AT_SEQUENCE_NUMBER',
+            StreamArn: message.DDBStreamBatchInfo.streamArn,
+            SequenceNumber: message.DDBStreamBatchInfo.startSequenceNumber,
+        })
+        .promise()
+        .catch((e) => {
+            throw new Error(`Failed to get shard iterator: ${JSON.stringify(e)}`);
+        });
+
+    const records = await dynamodbstreams
+        .getRecords({
+            ShardIterator: resp.ShardIterator,
+        })
+        .promise()
+        .catch((e) => {
+            throw new Error(`Failed to get records: ${JSON.stringify(e)}`);
+        });
+
+    logger.debug('Fetched records', JSON.stringify(records));
+    return records;
+};
+
+/**
+ * This Lambda function will read message from the DDbToEs DLQ, fetch records from the DynamoDB stream
+ * referenced in the queued message, and invoke the DDBToEs Lambda function providing the records as input.
+ *
+ * Note:
+ *   1. Message retention period is 14 days. After retention period, messages would be deleted automatically.
+ *   2. Record retention period is 24 hours. After retention period, records would be deleted automatically.
+ *   3. Message visibility timeout must be greater than the total time of processing the message,
+ *      otherwise the message would be received by this Lambda function again. Too long timeout would
+ *      make the messages invisible for long time, and it appears there are no messages in DLQ.
+ */
+exports.handler = async (event) => {
```

Comment:
This lambda only seems to work the SQS DLQ. How would we handle resyncing DDB documents that are not in the DLQ? Would we need another process to write DLQ messages?

We will have scenarios where ES becomes corrupted or out of sync besides transient failures in the ddbToES lambda.

Reply:
This lambda handles DQL messages only so far. It seems if we want to resync DDB documents, we need another lambda function.

Reply:
Correct. This function's purpose is only to work the SQS queue. The codebase could serve both functions, but we need to better understand how to rebuild a corrupted index.